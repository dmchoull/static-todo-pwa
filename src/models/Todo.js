import PropTypes from 'prop-types';

const Todo = (text, complete = false) => Object.freeze({ text, complete });

Todo.propTypes = PropTypes.shape({
  text: PropTypes.string.isRequired,
  complete: PropTypes.bool,
});

export default Todo;
