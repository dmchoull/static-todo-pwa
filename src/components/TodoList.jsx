import React from 'react';
import PropTypes from 'prop-types';
import Todo from '../models/Todo';
import TodoItem from '../containers/TodoItem';
import styles from '../assets/stylesheets/todo_list.pcss';

export default class TodoList extends React.Component {
  static propTypes = {
    todos: PropTypes.arrayOf(Todo.propTypes),
    addTodo: PropTypes.func.isRequired,
  };

  static defaultProps = {
    todos: [],
  };

  handleKeyDown = (e) => {
    if (e.keyCode !== 13) {
      return;
    }

    const input = e.target;
    this.props.addTodo(Todo(input.value));
    input.value = '';
  };

  render() {
    const { todos } = this.props;
    return (
      <div className={styles.todoapp}>
        <header className={styles.header}>
          <h1>todos</h1>
          <input
            className={styles.newTodo}
            data-testid="add-todo"
            type="text"
            placeholder="What needs to be done?"
            onKeyDown={this.handleKeyDown}
          />
        </header>

        <section className={styles.main}>
          <ul data-testid="todo-list" className={styles.todoList}>
            {todos.map(todo => <TodoItem key={todo.text} todo={todo} />)}
          </ul>
        </section>
      </div>
    );
  }
}
