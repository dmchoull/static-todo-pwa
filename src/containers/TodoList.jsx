import { connect } from 'react-redux';
import TodoList from '../components/TodoList';
import { addTodo } from '../actions/todo';

const mapStateToProps = ({ todos }) => ({
  todos,
});

const mapDispatchToProps = dispatch => ({
  addTodo: todo => dispatch(addTodo(todo)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);
