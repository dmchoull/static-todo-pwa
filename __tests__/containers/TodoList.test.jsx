import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { render, Simulate } from 'react-testing-library';
import 'dom-testing-library/extend-expect';
import TodoList from '../../src/containers/TodoList';
import reducer from '../../src/reducers';
import Todo from '../../src/models/Todo';

function renderWithRedux(ui, { initialState, store = createStore(reducer, initialState) } = {}) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    store,
  };
}

describe('TodoList', () => {
  test('renders todos', () => {
    const initialState = { todos: [Todo('item 1'), Todo('item 2', true)] };
    const { queryByText } = renderWithRedux(<TodoList />, { initialState });

    expect(queryByText('item 1')).toBeInTheDOM();
    expect(queryByText('item 2')).toBeInTheDOM();
    expect(queryByText('item 3')).not.toBeInTheDOM();
  });

  test('adding a new item', () => {
    const initialState = { todos: [Todo('item 1')] };
    const { queryByText, getByTestId } = renderWithRedux(<TodoList />, { initialState });

    const input = getByTestId('add-todo');
    input.value = 'item 2';
    Simulate.change(input);
    Simulate.keyDown(input, { key: 'Enter', keyCode: 13 });

    expect(queryByText('item 1')).toBeInTheDOM();
    expect(queryByText('item 2')).toBeInTheDOM();
    expect(input.value).toEqual('');
  });

  test('completing a todo', () => {
    const initialState = { todos: [Todo('item 1'), Todo('item 2')] };
    const { getByTestId, store } = renderWithRedux(<TodoList />, { initialState });

    const checkbox = getByTestId('complete-item 2');
    checkbox.value = true;
    Simulate.change(checkbox);

    expect(store.getState().todos[1].complete).toBeTruthy();
    expect(getByTestId('todo-list')).toMatchSnapshot();
  });

  test('un-completing a completed todo', () => {
    const initialState = { todos: [Todo('item 1'), Todo('item 2', true)] };
    const { getByTestId, store } = renderWithRedux(<TodoList />, { initialState });

    const checkbox = getByTestId('complete-item 2');
    checkbox.value = false;
    Simulate.change(checkbox);

    expect(store.getState().todos[1].complete).toBeFalsy();
    expect(getByTestId('todo-list')).toMatchSnapshot();
  });
});
