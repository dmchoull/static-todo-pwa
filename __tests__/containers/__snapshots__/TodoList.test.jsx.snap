// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList completing a todo 1`] = `
<ul
  class="todoList"
  data-testid="todo-list"
>
  <li
    class="todoItem "
  >
    <input
      class="toggle"
      data-testid="complete-item 1"
      type="checkbox"
    />
    <label>
      item 1
    </label>
  </li>
  <li
    class="todoItem completed"
  >
    <input
      class="toggle"
      data-testid="complete-item 2"
      type="checkbox"
      value="true"
    />
    <label>
      item 2
    </label>
  </li>
</ul>
`;

exports[`TodoList un-completing a completed todo 1`] = `
<ul
  class="todoList"
  data-testid="todo-list"
>
  <li
    class="todoItem "
  >
    <input
      class="toggle"
      data-testid="complete-item 1"
      type="checkbox"
    />
    <label>
      item 1
    </label>
  </li>
  <li
    class="todoItem "
  >
    <input
      class="toggle"
      data-testid="complete-item 2"
      type="checkbox"
      value="false"
    />
    <label>
      item 2
    </label>
  </li>
</ul>
`;
